### PR TITLE
Switch to `esbuild` for faster builds and also to allow using more modern JavaScript syntax and features

### PR DESCRIPTION
**Pull request by archmoj**
@plotly/plotly_js 

TODOs:
- [x] Fix `regl_codegen`
- [x] Fix custom bundles
- [x] Test new bundle in `amdefine`
- [x] Test new bundle in `requirejs`
- [x] Test new bundle in `kaleido`
- [x] Test new bundle in `orca`
- [x] additional QA on the python side on loading the bundle

### CODE REVIEW

**Comment by birkskyum** [comment]
Just want to point out that RSPack is in v1.0.0 alpha.1 ([link](https://www.rspack.dev/blog/announcing-1-0-alpha)) giving a path forward for webpack projects virtually without code changes.

**Comment by archmoj** [comment]
@birkskyum,
There are now conflicts on this branch after merging your recent PRs.
BTW if I recall karma is adjusted with esbuild.

**Comment by birkskyum** [comment]
I've tried rebasing, but it's quite hard. I think the easiest way to get this PR going again, is to make a new branch off master and apply the needed changes from this PR.

**Comment by archmoj** [comment]
> I've tried rebasing, but it's quite hard. I think the easiest way to get this PR going again, is to make a new branch off master and apply the needed changes from this PR.

Thanks for the note.
Let me try fixing it now.

**Comment by archmoj** [comment]
@birkskyum simply FYI: I noticed an build problem requiring map arcgis styles that were in json format.
I fixed it but converting then to js files in [8ba1116](https://github.com/plotly/plotly.js/pull/6909/commits/8ba111610bd50f0b9a93b3a0f6c64f5d66f8914f).

Then only notable failing test we should care about right now is the `requirejs` test.
That's how `plotly.js` is imported in `plotly.py` which is blocking this PR.
I thought some of the updates you made e.g. `jsdom` may help find a fix for this problem.
Perhaps there is a problem with UMD header, or arrow functions, or maybe we could make use of [requirejs-esm](https://www.npmjs.com/package/requirejs-esm)

**Comment by birkskyum** [comment]
If the requirejs is missing CI tests, how can I manually test it?

**Comment by archmoj** [comment]
> If the requirejs is missing CI tests, what's the step to manually test it?

@birkskyum It's not missing on CI.
For local testing you could run
npm run build && npm run test-requirejs

**Comment by birkskyum** [comment]
@archmoj , yes it's the amd module wrapper that's missing.

This can be verified by making a separate plotly.min.amd.js file, with this content in the dist folder:

```ts
define([], function() {
  const plotly = require('<absolute repo path>/dist/plotly.min');
  return plotly;
});
```

And loading that instead in test-requirejs:

```ts
requirejs.config({
    paths: {
        'plotly': '<absolute repo path>/dist/plotly.min.amd'
    }
});
```

Will make the test path

**Comment by birkskyum** [comment]
Configuring esbuild to make an `iife` and have the umdwrapper make `amd` like this:

```ts
module.exports = {
    entryPoints: ['./lib/index.js'],
    format: 'iife',
    globalName: 'Plotly',
    bundle: true,
    minify: false,
    sourcemap: false,
    plugins: [
        glsl({
            minify: true,
        }),
        environmentPlugin({
            NODE_DEBUG: false,
        }),
        umdWrapper({
            libraryName: 'Plotly',
            amd: true,
            cjs: true,
            global: true
        })
    ],
    alias: {
        stream: 'stream-browserify',
    },
    define: {
        global: 'window',
    },
    target: 'es2016',
    logLevel: 'info',
};

```

Gives the outcome that requirejs can load the plotly like this:

<details>
<summary>
output from test-requirejs
</summary>

```sh
Plotly loaded successfully: {
  version: '3.8.2',
  ascending: [Function: f],
  descending: [Function (anonymous)],
  min: [Function (anonymous)],
  max: [Function (anonymous)],
  extent: [Function (anonymous)],
  sum: [Function (anonymous)],
  mean: [Function (anonymous)],
  quantile: [Function (anonymous)],
  median: [Function (anonymous)],
  variance: [Function (anonymous)],
  deviation: [Function (anonymous)],
  bisectLeft: [Function: left],
  bisectRight: [Function: right],
  bisect: [Function: right],
  bisector: [Function (anonymous)],
  shuffle: [Function (anonymous)],
  permute: [Function (anonymous)],
  pairs: [Function (anonymous)],
  transpose: [Function (anonymous)],
  zip: [Function (anonymous)],
  keys: [Function (anonymous)],
  values: [Function (anonymous)],
  entries: [Function (anonymous)],
  merge: [Function (anonymous)],
  range: [Function (anonymous)],
  map: [Function (anonymous)],
  nest: [Function (anonymous)],
  set: [Function (anonymous)],
  behavior: { drag: [Function (anonymous)], zoom: [Function (anonymous)] },
  rebind: [Function (anonymous)],
  dispatch: [Function (anonymous)],
  event: null,
  requote: [Function (anonymous)],
  selection: [Function (anonymous)] { enter: [Function: Et] },
  ns: {
    prefix: {
      svg: 'http://www.w3.org/2000/svg',
      xhtml: 'http://www.w3.org/1999/xhtml',
      xlink: 'http://www.w3.org/1999/xlink',
      xml: 'http://www.w3.org/XML/1998/namespace',
      xmlns: 'http://www.w3.org/2000/xmlns/'
    },
    qualify: [Function: qualify]
  },
  select: [Function (anonymous)],
  selectAll: [Function (anonymous)],
  mouse: [Function (anonymous)],
  touch: [Function (anonymous)],
  touches: [Function (anonymous)],
  interpolateZoom: [Function (anonymous)],
  color: [Function: Jr],
  hsl: [Function: Ke],
  hcl: [Function: yt],
  lab: [Function: Vr],
  rgb: [Function: mi],
  functor: [Function: ai],
  xhr: [Function (anonymous)],
  dsv: [Function (anonymous)],
  csv: [Function: Fe] {
    parse: [Function (anonymous)],
    parseRows: [Function (anonymous)],
    format: [Function (anonymous)],
    formatRows: [Function (anonymous)]
  },
  tsv: [Function: Fe] {
    parse: [Function (anonymous)],
    parseRows: [Function (anonymous)],
    format: [Function (anonymous)],
    formatRows: [Function (anonymous)]
  },
  timer: [Function (anonymous)] { flush: [Function (anonymous)] },
  round: [Function (anonymous)],
  geom: {
    hull: [Function (anonymous)],
    polygon: [Function (anonymous)],
    voronoi: [Function (anonymous)],
    delaunay: [Function (anonymous)],
    quadtree: [Function (anonymous)]
  },
  interpolateRgb: [Function: al],
  interpolateObject: [Function: Ns],
  interpolateNumber: [Function: rl],
  interpolateString: [Function: Do],
  interpolate: [Function: iu],
  interpolators: [ [Function (anonymous)] ],
  interpolateArray: [Function: fu],
  ease: [Function (anonymous)],
  interpolateHcl: [Function: rv],
  interpolateHsl: [Function: ep],
  interpolateLab: [Function: Bh],
  interpolateRound: [Function: eh],
  transform: [Function (anonymous)],
  interpolateTransform: [Function: av],
  layout: {
    bundle: [Function (anonymous)],
    chord: [Function (anonymous)],
    force: [Function (anonymous)],
    hierarchy: [Function (anonymous)],
    partition: [Function (anonymous)],
    pie: [Function (anonymous)],
    stack: [Function (anonymous)],
    histogram: [Function (anonymous)],
    pack: [Function (anonymous)],
    tree: [Function (anonymous)],
    cluster: [Function (anonymous)],
    treemap: [Function (anonymous)]
  },
  random: {
    normal: [Function: normal],
    logNormal: [Function: logNormal],
    bates: [Function: bates],
    irwinHall: [Function: irwinHall]
  },
  scale: {
    linear: [Function (anonymous)],
    log: [Function (anonymous)],
    pow: [Function (anonymous)],
    sqrt: [Function (anonymous)],
    ordinal: [Function (anonymous)],
    category10: [Function (anonymous)],
    category20: [Function (anonymous)],
    category20b: [Function (anonymous)],
    category20c: [Function (anonymous)],
    quantile: [Function (anonymous)],
    quantize: [Function (anonymous)],
    threshold: [Function (anonymous)],
    identity: [Function (anonymous)]
  },
  svg: {
    arc: [Function (anonymous)],
    line: [Function (anonymous)] { radial: [Function (anonymous)] },
    area: [Function (anonymous)] { radial: [Function (anonymous)] },
    chord: [Function (anonymous)],
    diagonal: [Function (anonymous)] { radial: [Function (anonymous)] },
    symbol: [Function (anonymous)],
    symbolTypes: [
      'circle',
      'cross',
      'diamond',
      'square',
      'triangle-down',
      'triangle-up'
    ],
    axis: [Function (anonymous)],
    brush: [Function (anonymous)]
  },
  transition: [Function (anonymous)],
  text: [Function (anonymous)],
  json: [Function (anonymous)],
  html: [Function (anonymous)],
  xml: [Function (anonymous)]
}
```
</details>

.. but it doesn't have a PlotSchema property on it. Where is that set?

**Comment by archmoj** [comment]
@birkskyum That object looks like the d3 object not Plotly. The amd header was present in the build last time I checked it.
I suspect this might be related to a PR we accepted in v2 where the this keyword was replaced with self in our d3 fork to make our bundle work with es import.

**Comment by birkskyum** [comment]
There is in `preprocess` this section:

```ts
exposePartsInLib()
function exposePartsInLib() {
    var obj = {};

    var insert = function(name, folder) {
        obj[name] = folder + '/' + name;
    };

    insert('core', 'src');
 ...
 }
 ```
 
 And when looking at the `core` that is inserted it has:
 
 ```ts
 ...
 
 exports.Snapshot = require('./snapshot');
exports.PlotSchema = require('./plot_api/plot_schema');

 ```

**Comment by archmoj** [comment]
> @birkskyum That object looks like the d3 object not Plotly. The amd header was present in the build last time I checked it. I suspect this might be related to a PR we accepted in v2 where the this keyword was replaced with self in our d3 fork to make our bundle work with es import.

One of these PRs:
https://github.com/plotly/d3/pulls?q=is%3Apr+is%3Aclosed

Does it work if one loads last d3 v3 from original repo?

**Comment by birkskyum** [comment]
If I install `"d3": "^3.5.17",` and run `npm run build` I get:

```sh
    at src/core.js (about:blank:68157:7)
    at __require (about:blank:34:52)
    at lib/core.js (about:blank:68226:24)
file:///Users/admin/repos/plotly.js/tasks/schema.mjs:56
    var obj = Plotly.PlotSchema.get();
                     ^

TypeError: Cannot read properties of undefined (reading 'PlotSchema')
    at makeSchema (file:///Users/admin/repos/plotly.js/tasks/schema.mjs:56:22)
    at file:///Users/admin/repos/plotly.js/tasks/schema.mjs:85:1
    at ModuleJob.run (node:internal/modules/esm/module_job:195:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:337:24)
    at async loadESM (node:internal/process/esm_loader:34:7)
    at async handleMainPromise (node:internal/modules/run_main:106:12)
```

**Comment by archmoj** [comment]
How about the intermediate @plotly/d3 versions e.g. 3.8.0?

**Comment by birkskyum** [comment]
That gives the same as the latest plotly/d3 [here](https://github.com/plotly/plotly.js/pull/6909#issuecomment-2298533108)

File: src/components/drawing/symbol_defs.js
Comment (birkskyum) [inline src/components/drawing/symbol_defs.js:4]: out of curiosity, how is this related to esbuild?

**Comment by birkskyum** [comment]
Hmm, the `npm run test-jasmine -- surface` passes 17 tests locally, so I wonder why they fail in ci

File: package.json
Comment (birkskyum) [inline package.json:127]: where is `assert` used?

File: src/components/drawing/symbol_defs.js
Comment (archmoj) [inline src/components/drawing/symbol_defs.js:4]: @birkskyum I noticed requirejs somehow get confused with this function we added to our fork in https://github.com/d3/d3/commit/ddf0fe85f0d30887637bf73881f6a5ce37b4644e 
So I decided to simply inline it here.
I tried to revert the change once the all the `publish-dist` tests pass; but it didn't work. See https://app.circleci.com/pipelines/github/plotly/plotly.js/11179/workflows/d58363bf-53e6-48b5-bf0f-b3fb77e03570/jobs/247284/parallel-runs/0/steps/0-115
So I apply this change again. And it solved our problem.

File: package.json
Comment (archmoj) [reply package.json]: Good eye. It might be the result of a mistake when resolving the merge conflicts.
I uninstalled it in https://github.com/plotly/plotly.js/pull/6909/commits/9f0267b815ac7f1a09fb4da331e752822dd71254.

**Comment by archmoj** [comment]
> The `npm run test-jasmine -- surface` and `parcoords` appear to pass everything locally, so I wonder why they fail in ci

@birkskyum OK. Something strange seems to start happening today.
Those jasmine tests pass when I merged your PR. 
But now if I rerun they fail. 
This is possibly related to latest Chrome version if we didn't lock it? 
To possibly confirm that scenario I'm rerunning the tests on master now.

**Comment by archmoj** [comment]
> > The `npm run test-jasmine -- surface` and `parcoords` appear to pass everything locally, so I wonder why they fail in ci
> 
> @birkskyum OK. Something strange seems to start happening today. Those jasmine tests pass when I merged your PR. But now if I rerun they fail. This is possibly related to latest Chrome version if we didn't lock it? To possibly confirm that scenario I'm rerunning the tests on master now.

@birkskyum Yes it failed on master too. So the Chrome version should be pinned on master.

**Comment by birkskyum** [comment]
This PR takes my local build time from 3m 24.3s down to 12.1s, so going back to the main branch already feels like building in slow motion - well done @archmoj !

**Comment by gvwilson** [comment]
17X performance improvement is exciting.

**Comment by archmoj** [comment]
One could argue it's actually more than that. 
On CI if we look at CI-build task in isolation it used to take 90 seconds. Now it's only 3 seconds.
So 20-30X seems like a better number.

**Comment by archmoj** [comment]
@alexcjohnson @gvwilson 
This PR is ready to go :rocket:

**Comment by gvwilson** [comment]
thank you - @marthacryan can you please have a look at this as well?

File: .circleci/config.yml
Comment (marthacryan) [inline .circleci/config.yml:78]: Is this related to the changes in this PR?

File: CUSTOM_BUNDLE.md
Comment (marthacryan) [inline CUSTOM_BUNDLE.md:73]: I'm not against removing unminified here, but could you provide some context on why that's in this PR?

File: .circleci/config.yml
Comment (archmoj) [reply .circleci/config.yml:78]: Not really. Since other containers finish earlier, I made this adjustment so that this container also finishes around the same time and catches potential errors earlier.

File: CUSTOM_BUNDLE.md
Comment (archmoj) [reply CUSTOM_BUNDLE.md:73]: Good question. I need to change centralized partial bundle script to bring back these options. Having only minified bundle could save users few seconds. So I'll try to revert this.

Comment (archmoj) [reply CUSTOM_BUNDLE.md:73]: Reverted and addressed in [8bb92f8](https://github.com/plotly/plotly.js/pull/6909/commits/8bb92f8c02d7d3855dbf2c2d161a9e265d8e2d96).

**Comment by archmoj** [comment]
I'm looking forward for @alexcjohnson's review on this as well. Also thanks to @marthacryan and @birkskyum reviews & contributions :pray:

This PR helps unblock `Chart2Music keyboard and sound accessibility features` in #6680. cc: @Coding-with-Adam  

It might be a good idea and timing to include this in the upcoming RC release of [v2.35.0] (https://github.com/plotly/plotly.js/milestone/72) and our QA. cc: @gvwilson @ndrezn @LiamConnors 

It would also help our ongoing and future development and testing of other pull requests. cc: @emilykl @antoinerg

**Comment by gvwilson** [comment]
I approve. Let's hold off on the merge until next week (feels too big to do on a Friday), but congratulations everyone - well done.

**Comment by ndrezn** [comment]
Super excited about this! This change will really improve the development experience for Plotly.js. Kudos 🥳

**Comment by Coding-with-Adam** [comment]
Wonderful news 🚀 . Thank you @archmoj; Thank you @marthacryan

**Review by alexcjohnson — APPROVED** [review]
💃 I'm super excited for this. Looks great!

**Comment by archmoj** [comment]
Thanks for the reviews!
Before merging this PR, I'd like to do additional QA on the python side on loading the bundle.

**Comment by birkskyum** [comment]
We'll need to add esbuild-style-plugin too, because of the require(file.css) in: 
- https://github.com/plotly/plotly.js/pull/7140#discussion_r1747626991

**Pull request merged**
